### PR TITLE
Extract Mesh VonMises Stresses

### LINKED ISSUE #463: Allow to Extract Mesh VonMises Stresses
#### Description:
Currently Von Mises Shell stresses outputs are not retrievable via BHoM.
It looks that ETABS API has a function allowing to retrieve them though (cAnalysisResults.AreaStressShell Method).
It would be great to test them and, if working, to incorporate them in the BHoM.

Pull Von Mises Stresses Issue 01

Pull Von Mises Stresses Issue 02

### PR DESCRIPTION
**Pull request by GCRA101**
### Issues addressed by this PR

 Closes #463 

SvmBHoMPlots

SvmGHvsExcelTableCheck

ETABS Toolkit now can allow to extract Shell Von Mises Stress outputs from ETABS.

 ### Test files

Files Folder
https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23471-ExtractMeshVonMisesStresses?csf=1&web=1&e=OpP1iW

Grasshopper File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23471-ExtractMeshVonMisesStresses/Test%20Script.gh?csf=1&web=1&e=AFd4Si

ETABS File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23471-ExtractMeshVonMisesStresses/Test%20ETABS%20Model.EDB?csf=1&web=1&e=6fgDcc

Excel File
https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23471-ExtractMeshVonMisesStresses/Test%20ETABS%20Model.EDB?csf=1&web=1&e=6fgDcc

 ### Changelog

- Added methods ReadMeshVonMises, SmoothenVonMisesStresses and ComputeVonMisesMoment to read/compute Von Mises stresses from etabs shell objects.
- Added utility method GetPanelThickness for getting the thickness of etabs panels

### CODE REVIEW

**Comment by GCRA101** [comment]
@IsakNaslundBh, @Chrisshort92 , @peterjamesnugent, just to be clear, this Pull Request is for the next BHoM 7.4 Milestone in autumn. Not the one coming up now in september. ;)

**Review by Chrisshort92 — APPROVED** [review]
Functionality Tested with provided test files and working as expected. approved for merge

**Comment by Chrisshort92** [comment]
@BHoMBot

**Comment by peterjamesnugent** [comment]
@BHoMBot check versioning

**Comment by GCRA101** [comment]
@BHoMBot check copyright-compliance

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 1]
```diff
@@ -42,6 +42,7 @@
 using BH.Engine.Geometry;
 using BH.oM.Adapter;
 using BH.oM.Structure.Elements;
+using System.Xml.Linq;
```
Comment (peterjamesnugent) [inline Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:45]:
```suggestion
```
Don't need this

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 2]
```diff
@@ -243,8 +247,10 @@ private List<MeshResult> ReadMeshStress(List<string> panelIds, List<string> case
...
-                            MeshStress mStressBot = new MeshStress(panelIds[i], pointElm[j], elm[j], loadCase[j], mode, timeStep, MeshResultLayer.Lower, 0, MeshResultSmoothingType.None, oM.Geometry.Basis.XY, s11Bot[j], s22Bot[j], s12Bot[j], s13Avg[j], s23Avg[j], sMaxBot[j], sMinBot[j], double.NaN);
+                            MeshStress mStressTop = new MeshStress(panelIds[i], pointElm[j], elm[j], loadCase[j], mode, timeStep, MeshResultLayer.Upper, 1, MeshResultSmoothingType.None, 
+                                oM.Geometry.Basis.XY, s11Top[j], s22Top[j], s12Top[j], s13Avg[j], s23Avg[j], sMaxTop[j], sMinTop[j], double.NaN);
+                            MeshStress mStressBot = new MeshStress(panelIds[i], pointElm[j], elm[j], loadCase[j], mode, timeStep, MeshResultLayer.Lower, 0, MeshResultSmoothingType.None, 
+                                oM.Geometry.Basis.XY, s11Bot[j], s22Bot[j], s12Bot[j], s13Avg[j], s23Avg[j], sMaxBot[j], sMinBot[j], double.NaN);
```
Comment (peterjamesnugent) [inline Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:253]: Has anything actually changed between the code above and the \develop branch?

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 3]
```diff
@@ -271,6 +277,124 @@ private List<MeshResult> ReadMeshStress(List<string> panelIds, List<string> case
...
+                    return new List<MeshResult>();
+            }
+
+            if (layer == MeshResultLayer.Upper || layer == MeshResultLayer.Lower)
+            {
+                Engine.Base.Compute.RecordWarning("Results for both bot and top layers will be extracted at the same time");
+            }
+            else
+            {
+                Engine.Base.Compute.RecordWarning("Stress extraction is currently only possible at bot and top layers. Please update the MeshResultLayer parameter");
```
Comment (peterjamesnugent) [inline Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:299]:
```suggestion
                Engine.Base.Compute.RecordWarning("Stress extraction is currently only possible at bot and top layers. Please update the MeshResultLayer parameter.");
```

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 4]
```diff
@@ -271,6 +277,124 @@ private List<MeshResult> ReadMeshStress(List<string> panelIds, List<string> case
...
+            double[] s11Top = null, s22Top = null, s12Top = null, sMaxTop = null, sMinTop = null, sAngTop = null, svmTop = null;
+            double[] s11Bot = null, s22Bot = null, s12Bot = null, sMaxBot = null, sMinBot = null, sAngBot = null, svmBot = null;
+            double[] s13Avg = null, s23Avg = null, sMaxAvg = null, sAngAvg = null;
+            double[] f11 = null, f22 = null, f12 = null, fMax = null, fMin = null, fAngle = null, fvm = null;
+            double[] m11 = null, m22 = null, m12 = null, mMax = null, mMin = null, mAngle = null; 
+            double[] v13 = null, v23 = null, vMax = null, vAngle=null;
+
+            List<MeshResult> results = new List<MeshResult>();
+
+
```
Comment (peterjamesnugent) [inline Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:317]: Clean up whitespace

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 5]
```diff
@@ -271,6 +277,124 @@ private List<MeshResult> ReadMeshStress(List<string> panelIds, List<string> case
...
+            if (smoothing == MeshResultSmoothingType.ByPanel)
+                Engine.Base.Compute.RecordWarning("Stress values have been smoothed outside the API by averaging all force values in each node");
```
Comment (peterjamesnugent) [inline Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:319]:
```suggestion
                Engine.Base.Compute.RecordWarning("Stress values have been smoothed outside the API by averaging all force values in each node.");
```

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 6]
```diff
@@ -271,6 +277,124 @@ private List<MeshResult> ReadMeshStress(List<string> panelIds, List<string> case
...
+                    if ((ret1 == 0) && (ret2 == 0))
+                    {
+
+                        for (int j = 0; j < resultCount; j++)
+                        {
+                            int mode;
+                            double timeStep;
+
+                            // Calculate Von Mises Moment
+                            double Mvm = ComputeVonMisesMoment(svmTop[j], svmBot[j], panelThk);
```
Comment (peterjamesnugent) [inline Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:360]: Should we be calculating moments here? If ETABS doesn't output the vonMises moment, I think we should just state that in the documentation and then the user can calculate the Mvm as they see fit.

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 7]
```diff
@@ -481,7 +669,9 @@ private List<MeshForce> SmoothenForces(List<MeshForce> forces)
                 double vx = group.Average(x => x.VX);
                 double vy = group.Average(x => x.VY);
 
-                smoothenedForces.Add(new MeshForce(first.ObjectId, first.NodeId, "", first.ResultCase, first.ModeNumber, first.TimeStep, first.MeshResultLayer, first.LayerPosition, MeshResultSmoothingType.ByPanel, first.Orientation, nxx, nyy, nxy, mxx, myy, mxy, vx, vy));
+                smoothenedForces.Add(new MeshForce(first.ObjectId, first.NodeId, "", first.ResultCase, first.ModeNumber, first.TimeStep, 
+                                                        first.MeshResultLayer, first.LayerPosition, MeshResultSmoothingType.ByPanel, first.Orientation, 
+                                                        nxx, nyy, nxy, mxx, myy, mxy, vx, vy));
```
Comment (peterjamesnugent) [inline Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:674]: Why have these been modified? This is going to confuse people who look back on this in a few years wondering what changes you made to `MeshForce` either via the PR or Blame.

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 8]
```diff
@@ -508,14 +698,48 @@ private List<MeshStress> SmoothenStresses(List<MeshStress> forces)
                 double pr2 = group.Average(x => x.Principal_2);
                 double pr1_2 = group.Average(x => x.Principal_1_2);
 
-                smoothenedForces.Add(new MeshStress(first.ObjectId, first.NodeId, "", first.ResultCase, first.ModeNumber, first.TimeStep, first.MeshResultLayer, first.LayerPosition, MeshResultSmoothingType.ByPanel, first.Orientation, sxx, syy, sxy, txx, tyy, pr1, pr2, pr1_2));
+                smoothenedForces.Add(new MeshStress(first.ObjectId, first.NodeId, "", first.ResultCase, first.ModeNumber, first.TimeStep, 
+                                                        first.MeshResultLayer, first.LayerPosition, MeshResultSmoothingType.ByPanel, first.Orientation, 
+                                                        sxx, syy, sxy, txx, tyy, pr1, pr2, pr1_2));
```
Comment (peterjamesnugent) [inline Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:703]: Why have these been modified? This is going to confuse people who look back on this in a few years wondering what changes you made to `MeshForce` either via the PR or Blame.

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 9]
```diff
@@ -508,14 +698,48 @@ private List<MeshStress> SmoothenStresses(List<MeshStress> forces)
...
+        {
+            List<MeshVonMises> smoothenedVMStresses = new List<MeshVonMises>();
+
+            foreach (IEnumerable<MeshVonMises> group in forces.GroupBy(x => new { x.ResultCase, x.TimeStep, x.NodeId }))
+            {
+                MeshVonMises first = group.First();
+
+                double s = group.Average(x => x.S);
+                double n = group.Average(x => x.N);
+                double m = group.Average(x => x.M);
```
Comment (peterjamesnugent) [inline Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:722]: If you remove the vonMises moment as commented above, this is going to be an average of `double.NaN` - you may need to add some checks in here.

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 10]
```diff
@@ -508,14 +698,48 @@ private List<MeshStress> SmoothenStresses(List<MeshStress> forces)
...
+
+        /***************************************************/
+
+        private double ComputeVonMisesMoment(double svmTop, double svmBot, double thk)
+        {
+            double svmAvg = (svmTop + svmBot) / 2;
+            double vonMisesMoment = ((svmBot - svmAvg) * (thk / 2) * (1 / 2) * (thk - 2 * thk / 2 * 1 / 3)) / 1000;
+
+            return vonMisesMoment;
+        }
```
Comment (peterjamesnugent) [inline Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:739]:
```suggestion
```
As above

File: Etabs_Adapter/ETABSAdapter.cs [hunk 11]
```diff
@@ -79,6 +79,7 @@ public ETABS17Adapter(string filePath = "", EtabsSettings etabsSetting = null, b
         [Description("Creates an adapter to ETABS version 18 or later. For connection to ETABS v17 use the ETABS17Adapter. For connection to ETABS 2016 use the ETABS2016Adapter. Earlier versions not supported.")]
         public ETABSAdapter(string filePath = "", EtabsSettings etabsSetting = null, bool active = false)
 #endif
+
```
Comment (peterjamesnugent) [inline Etabs_Adapter/ETABSAdapter.cs:82]: Why is this file being edited? Please remove.

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]
I don't think we should be calculating the vonMises moment, everything is there for the user to do it themselves - if ETABS doesn't allow us to pull it, we shoudn't calculated it for the user.

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs (hunk 4 above)
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Read/Results/MeshResults.cs]: @peterjamesnugent, thanks for the comment.
Cleaned up and pushed to origin ;)

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs (hunk 5 above)
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Read/Results/MeshResults.cs]: @peterjamesnugent, suggestion committed ;)

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs (hunk 10 above)
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:697]: @peterjamesnugent, code updated in line with comment

File: Etabs_Adapter/ETABSAdapter.cs (hunk 11 above)
Comment (GCRA101) [reply Etabs_Adapter/ETABSAdapter.cs:82]: @peterjamesnugent blank line removed ;)

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs (hunk 7 above)
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:633]: @peterjamesnugent, looks to me here just the indentation of the code has changed. Would you prefer me to turn code lines alignment to their original?

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs (hunk 8 above)
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Read/Results/MeshResults.cs:662]: @peterjamesnugent, looks to me here just the indentation of the code has changed. Would you prefer me to turn code lines alignment to their original?

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs (hunk 7 above)
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Read/Results/MeshResults.cs]: @peterjamesnugent, code lines restored as per they original setting out ;)

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs (hunk 8 above)
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Read/Results/MeshResults.cs]: @peterjamesnugent, code lines restored as per they original setting out ;)

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs (hunk 9 above)
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Read/Results/MeshResults.cs]: @peterjamesnugent , removed the von mises moment. x.M=NaN in all instances now ;)

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]
Once you've removed the two suggestions, I am happy from a code perspective.

Can @Chrisshort92 review from a functionality perspective please.

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 12]
```diff
@@ -267,8 +260,113 @@ private List<MeshResult> ReadMeshStress(List<string> panelIds, List<string> case
...
+                    
+                    // Extract Von Mises Resultant Axial Forces
+                    ret2 = m_model.Results.AreaForceShell(panelIds[i], itemTypeElm, ref resultCount, 
+                        ref obj, ref elm, ref pointElm, ref loadCase, ref stepType, ref stepNum,
+                        ref f11, ref f22, ref f12, ref fMax, ref fMin, ref fAngle, ref fvm, 
+                        ref m11, ref m22, ref m12, ref mMax, ref mMin, ref mAngle, 
+                        ref v13, ref v23, ref vMax, ref vAngle);
+
+                    // Get the panel thickness
+                    panelThk = GetPanelThickness(panelIds[i]);
```
[the remark this answers is not in the document]
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Read/Results/MeshResults.cs]: @peterjamesnugent, thanks for the comment. Changes incorporated in latest commit.

File: Etabs_Adapter/CRUD/Read/Results/MeshResults.cs [hunk 13]
```diff
@@ -462,6 +543,65 @@ private List<string> CheckGetPanelIds(MeshResultRequest request)
...
+            // Case 7 - Slab
+            if (m_model.PropArea.GetSlab(areaPropName, ref slabType, ref shellType, ref matProp, ref slabThickness, ref color, ref notes, ref guid) == 0) return Math.Round(slabThickness, 3);
+            // Case 8 - Wall
+            if (m_model.PropArea.GetWall(areaPropName, ref wallPropType, ref shellType, ref matProp, ref thickness, ref color, ref notes, ref guid) == 0) return Math.Round(thickness, 3);
+
+            // Case else - No Thickness found...
+            return 0.0;
+
+
+        }
```
[the remark this answers is not in the document]
Comment (GCRA101) [reply Etabs_Adapter/CRUD/Read/Results/MeshResults.cs]: @peterjamesnugent, thanks for the comment. Changes incorporated in latest commit.

**Comment by peterjamesnugent** [comment]
@BHoMBot check required
@BHoMBot check copyright-compliance

**Review by HugoVanLooveren — APPROVED** [review]
Tested with BHoM Beta 7.2.1
Tested with ETABS 21

Despite not having used the von Mises in ETABS before, it worked fine. Checked the values in ETABS table with the numerical outputs.
The visible mesh output also was the same when displayed in ETABS (using max/min values)

So all good :)

**Comment by peterjamesnugent** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**